src/server/Server.java

```diff
@@ -9,92 +15,72 @@
-        MaxRequestSizeKB = in;
-    }
 
-    public void setBacklog(int in) {
-        backlog = in;
+    public Server(Function<HttpRequest, HttpResponse> handler) {
+        this(8080, false, handler);
     }
 
-    public void setMIMEFile(String in) throws Exception {
-        MIMEFile = in;
-        loadMIME();
+    public Server(int port, boolean useGzip, Function<HttpRequest, HttpResponse> handler) {
+        this(port, 100000, 4096, 4096, useGzip, handler);
     }
 
-    public void setWWWDirectory(String in) {
-        WWWDir = in;
+    public Server(int port, int maxConcurrentRequests, int maxRequestSizeBytes, int maxResponseSizeBytes, boolean useGzip, Function<HttpRequest, HttpResponse> handler) {
+        this(port, maxConcurrentRequests, maxRequestSizeBytes, useGzip, "www", "etc/MIME.db", handler);
     }
 
-    public void setPort(int in) {
-        Port = in;
+    public Server(int port, int maxConcurrentRequests, int maxRequestSizeBytes, boolean useGzip, String rootDir, String mimeFile, Function<HttpRequest, HttpResponse> handler) {
+        this.port = port;
+        this.maxConcurrentRequests = maxConcurrentRequests;
+        this.backlog = maxConcurrentRequests * 5;
+        this.maxRequestSizeBytes = maxRequestSizeBytes;
+        this.useGzip = useGzip;
+        this.rootDir = rootDir;
+        this.mimeFile = mimeFile;
+        this.handler = handler;
     }
 
-    public void addCustomHeader(Entry<String, String> in) {
-        CustomHeaders.put(in.getKey(), in.getValue());
+    private static String getStackTrace(final Throwable throwable) {
+        final StringWriter sw = new StringWriter();
+        final PrintWriter pw = new PrintWriter(sw, true);
+        throwable.printStackTrace(pw);
+        return sw.getBuffer().toString();
     }
 
     private void mainLoop(ServerSocket serverSocket, ThreadPoolExecutor poolExecutor) {
         while (true) {
             try {
                 Socket s = serverSocket.accept();
-                s.setKeepAlive(false);
+                s.setKeepAlive(true);
                 s.setTcpNoDelay(true);
-                s.setReceiveBufferSize(BufferSize);
-                s.setSendBufferSize(BufferSize);
-                s.setSoTimeout(SOTimeout);
+                s.setReceiveBufferSize(maxRequestSizeBytes);
+                s.setSendBufferSize(maxRequestSizeBytes);
+                s.setSoTimeout(60000);
 
                 try {
                     poolExecutor.execute(new Engine(s));
                 } catch (RejectedExecutionException ignore) {
                     log.i("concurrent connections exceed the configured maximum");
-                    Network.write(new BufferedOutputStream(s.getOutputStream()), new byte[]{}, "", HTTPCode.SERVICE_UNAVAILABLE, false, new HashMap<>(), false);
+                    Network.write(new BufferedOutputStream(s.getOutputStream()), new HttpResponse("", HttpStatusCode.SERVICE_UNAVAILABLE, HttpContentType.TEXT_PLAIN), false);
```

Comment:
HttpContentType.PLAIN_TEXT is more intuitive

Reply:
You're right but I wrote it to follow the pattern defined by the standard `type/subtype => text/plain => TEXT_PLAIN`

Reply:
ContentTypes should be acquired in MIME.db as well

Reply:
I think it's unnecessary, the interface is good for now.
We can modify it in a different version.